NETOBSERV-2106: eBPF Flows filtering blog (#8)
netobserv/netobserv.github.io: content/posts/2025-01-30-ebpf-flows-filtering.md

          ports: 443,6443
          protocol: TCP
          sampling: 10
        - action: Accept
          cidr: 10.129.0.1/24
          ports: 53
          protocol: UDP
          sampling: 20
        - action: Reject
          tcpFlags: "SYN"
          cidr: 10.130.0.0/24
          protocol: TCP
          sourcePorts: 80-100
        - action: Accept
          cidr: 172.30.0.0/16
          protocol: SCTP
          pktDrops: true
        - action: Reject
          cidr: 8.8.8.8/32
          protocol: ICMP
          icmpType: 8 // ICMP Echo request packet
```
2️⃣ Packet Inspection – Extract relevant packet attributes within an eBPF program.

3️⃣ Early Flow Filtering – Drop or pass packets based on predefined conditions before sending data to user space.

The Following Diagram shows how eBPF filtering is done

<img src="{page.image('ebpf-flows-filtering/ebpf-flows-filtering.png')}" alt="eBPF Flows Filtering Processing">

Note:
- The maximum number of eBPF flow filter rules is limited to 16 to ensure conservative memory
usage when this feature is enabled.

- rules with duplicate CIDRs are rejected via validation webhook.

- Both IPv4 and IPv6 formats are supported.

> - In case users wanted to match on any CIDR Null CIDR can be used like `0.0.0.0/0` which is also the default.

Reviewer: Line can be updated to `In case users wanted to match on any CIDR,  default Null CIDR `0.0.0.0/0` can also be used.`

## Key Use Cases

🚀 Reducing Observability Overhead – Avoid logging irrelevant flows in high-traffic Kubernetes clusters.

🔐 Security Filtering – Focus on anomalous or suspicious traffic while ignoring normal flows.

🌐 Network Performance Monitoring – Capture only high-latency or dropped-packet flows for troubleshooting.

### Filter EastWest and NorthSouth flows

Let's allow serviceIP 172.30.100.64:80 with sample 10
and allow pods traffic between subnet 10.128.0.0/16 and 10.129.0.0/16 with sampling 20 and reject everything else

```yaml
agent:
    type: eBPF
    ebpf:
      flowFilter:
        enable: true
        rules:
        - action: Accept
          cidr: 172.30.0.0/16
          sampling: 10
        - action: Accept
          cidr: 10.128.0.0/16
          peerCIDR: 10.129.0.0/16
          sampling: 20
        - action: Reject
          cidr: 0.0.0.0/0
``` 

<img src="{page.image('ebpf-flows-filtering/ebpf-svc-and-pods-flows.png')}" alt="eBPF Flows Filtering Kubernetes NorthSouth and EastWest Flows">

### Filter flows with packet drops

> let's filter any kubernetes service flows with a packet drop and reject everything else

Reviewer: Add Note for this saying PacketDrop feature needs to be enabled with eBPF in privileged mode

          cidr: 8.8.8.8/32
          protocol: ICMP
          icmpType: 8 // ICMP Echo request packet
```
2️⃣ Packet Inspection – Extract relevant packet attributes within an eBPF program.

3️⃣ Early Flow Filtering –
Skip or Allow packets based on predefined conditions before doing further ebpf packets processing.
The Following Diagram shows how eBPF filtering is done

<img src="{page.image('ebpf-flows-filtering/ebpf-flows-filtering.png')}" alt="eBPF Flows Filtering Processing">

Note:
- The maximum number of eBPF flow filter rules is limited to 16 to ensure conservative memory
usage when this feature is enabled.

- rules with duplicate CIDRs are rejected via validation webhook.

- Both IPv4 and IPv6 formats are supported.

- In case users wanted to match on any CIDR, default Null CIDR `0.0.0.0/0` can be used.

## Key Use Cases

🚀 Reducing Observability Overhead – Avoid logging irrelevant flows in high-traffic Kubernetes clusters.

🔐 Security Filtering – Focus on anomalous or suspicious traffic while ignoring normal flows.

🌐 Network Performance Monitoring – Capture only high-latency or dropped-packet flows for troubleshooting.

### Filter EastWest and NorthSouth flows

> Let's allow serviceIP 172.30.100.64:80 with sample 10

Reviewer: sampling 10

and allow pods traffic between subnet 10.128.0.0/16 and 10.129.0.0/16 with sampling 20 and reject everything else

```yaml
agent:
    type: eBPF
    ebpf:
      flowFilter:
        enable: true
        rules:
        - action: Accept
          cidr: 172.30.0.0/16
          sampling: 10
        - action: Accept
          cidr: 10.128.0.0/16
          peerCIDR: 10.129.0.0/16
          sampling: 20
        - action: Reject
          cidr: 0.0.0.0/0
``` 

<img src="{page.image('ebpf-flows-filtering/ebpf-svc-and-pods-flows.png')}" alt="eBPF Flows Filtering Kubernetes NorthSouth and EastWest Flows">

### Filter flows with packet drops

let's filter any kubernetes service flows with a packet drop and reject everything else,
please note for this use case to work need to enable `PacketDrop` feature which requires `privileged` flag

> to be set to `true` as shows in the following sample configuration

Reviewer: Can be updated to `Please not for this use case, we need to enable PacketDrop feature with eBPF in privileged mode as shown in the below configuration`

```yaml
agent:
    type: eBPF
    ebpf:
      privileged: true
      features:
      - PacketDrop
      flowFilter:
        enable: true
        rules:
        - action: Accept
          cidr: 172.30.0.0/16
          pktDrops: true
        - action: Reject
          cidr: 0.0.0.0/0
```

<img src="{page.image('ebpf-flows-filtering/ebpf-filter-svc-pkt-drops.png')}" alt="eBPF Flows Filtering Kubernetes Services with Packet Drop">

### Filter TCP flows with TCP Flags detect TCP SYN Flood attack

Using TCP Flags filtering can help detect when your cluster is under TCP Syn Flood attack

```yaml
agent:
    type: eBPF
    ebpf:
      flowFilter:
        enable: true
        rules:
        - action: Accept
          cidr: 0.0.0.0/0
          protocol: TCP
          tcpFlags: "SYN"
          sampling: 1
```

<img src="{page.image('ebpf-flows-filtering/ebpf-filter-with-tcpflags.png')}" alt="eBPF Flows Filtering TCP flows using TCP flags">

### Filter DNS query over ports 53 and 5353 for both TCP and UDP

> 

Reviewer: This does not have a title, like the other examples

Author: added one

```yaml
agent:
    type: eBPF
    ebpf:
      features:

>         - DNSTracking

Reviewer: By the way we dont need to have DNS Tracking for this filter to work. So maybe can tell that if DNS Tracking feature is enabled then can view the DNS Latency info as well or just not have it at all. either way works for me

Author: wanted to make it more useful usecase for users hence adding DNSTracking to it

description: NetObserv eBPF Flows Filtering
tags: eBPF,Monitoring,Troubleshooting
authors: [msherif1234]
---

# Flow Filtering in eBPF: Optimizing Resource Usage by Selecting Critical Flows

## Introduction

In high-traffic environments, processing every network flow can be resource-intensive,
leading to CPU overload and excessive memory usage.
eBPF-based flow filtering solves this challenge by selecting only important flows,
reducing system strain while maintaining visibility.

## Why Flow Filtering?

The primary goal of flow filtering is resource efficiency.
Instead of capturing and analyzing every flow, filtering mechanisms allow us to:

✅ Reduce CPU & Memory Overhead – Process only relevant traffic, avoiding unnecessary computation.

✅ Optimize Storage Usage – Store only meaningful flow records, reducing disk and database load.

✅ Enhance Performance – Minimize packet processing latency and improve system responsiveness.

✅ Focus on Critical Traffic – Prioritize important flows for security, compliance, and performance monitoring.

## How Flow Filtering Works in eBPF
eBPF allows filtering flows at the source, avoiding costly user-space processing. This typically involves:

> 1️⃣ Defining Filtering Rules – Specify criteria such as source/destination IP, port, protocol, 
> or application metadata, The following table shows all possible filtering options and their default setting

Reviewer 2: ```suggestion
1️⃣ Defining Filtering Rules – Specify criteria such as source/destination IP, port, protocol, 
or application metadata. The following table shows all possible filtering options and their default setting:
```

| icmpType    | ICMP type to match on the flow                           | for example 8 or 13                                                     |           |
| icmpCode    | ICMP code to match on the flow                           | for example 0 or 1                                                      |           |
| peerIP      | Peer IP to match on the flow                             | for example 1.1.1.1 or 1::1                                             |           |
| peerCIDR    | Peer IP CIDR to match on the flow                        | for example 1.1.1.1/24 or 1::1/48                                       |           |
| pktDrops    | filter flows with packets drop                           | true, false                                                             |           |
| sampling    | sampling rate to use for filtered flows                  | for example 10 or 20 (any value >= 1)                                   |           |
| tcpFlags    | TCP flags to filter flows by                             | "SYN";"SYN-ACK";"ACK";"FIN";"RST";"URG";"ECE";"CWR";"FIN-ACK";"RST-ACK" |           |

Note:

> - A rule cannot include both ports and either sourcePorts or destPorts simultaneously.

Reviewer 2: ```suggestion
- A rule cannot include both ports and either `sourcePorts` or `destPorts` simultaneously.
```

Reviewer 2: Since this is a possible option, I suggest denoting it in backticks

- Supports both IPv4 and IPv6 formats.
- If users wish to match any CIDR, the default `0.0.0.0/0` (Null CIDR) can be used.

The example below demonstrates various filtering options using multi-rule filters.

```yaml
agent:
  type: eBPF
  ebpf:
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 10.128.0.0/24
          peerCIDR: 10.129.0.0/24
          ports: '443,6443'
          protocol: TCP
          sampling: 10
        - action: Accept
          cidr: 10.129.0.1/24
          ports: 53
          protocol: UDP
          sampling: 20
        - action: Reject
          tcpFlags: SYN
          cidr: 10.130.0.0/24
          protocol: TCP
          sourcePorts: 80-100
        - action: Accept
          cidr: 172.30.0.0/16
          protocol: SCTP
          pktDrops: true
        - action: Reject
          cidr: 8.8.8.8/32
          protocol: ICMP
          icmpType: 8 // ICMP Echo request packet
```
2️⃣ Packet Inspection – Extract relevant packet attributes within an eBPF program.

3️⃣ Early Flow Filtering –
Skip or Allow packets based on predefined conditions before doing further ebpf packets processing.

> The Following Diagram shows how eBPF filtering is done

Reviewer 2: For the numbered list, I would suggest not using title case. Instead, use regular sentence case like this:
1. Defining filtering rules - 
2. Packet inspection - Extract relevant...
3. Early flow filtering - Skip or allow packets

> <img src="{page.image('ebpf-flows-filtering/ebpf-flows-filtering.png')}" alt="eBPF Flows Filtering Processing">

Author: I added more details hopefully that won't cause confusion :)

## Key Use Cases

🚀 Reducing Observability Overhead – Avoid logging irrelevant flows in high-traffic Kubernetes clusters.

🔐 Security Filtering – Focus on anomalous or suspicious traffic while ignoring normal flows.

🌐 Network Performance Monitoring – Capture only high-latency or dropped-packet flows for troubleshooting.

### Filter EastWest and NorthSouth flows

> Allow traffic to service IP 172.30.100.64:80 with a sampling rate of 10.

Reviewer 3: ```suggestion
Allow traffic to service IP `172.30.100.64:80` with a sampling rate of `10`.
```

> Permit communication between pods in the 10.128.0.0/16 and 10.129.0.0/16 subnets with a sampling rate of 20.

Reviewer 3: ```suggestion
Permit communication between pods in the `10.128.0.0/16` and `10.129.0.0/16` subnets with a sampling rate of `20`.
```

Reject all other traffic.

```yaml
agent:
  type: eBPF
  ebpf:
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 172.30.0.0/16
          sampling: 10
        - action: Accept
          cidr: 10.128.0.0/16
          peerCIDR: 10.129.0.0/16

>           sampling: 20

Reviewer 4: the problem with that rule is that it wouldn't accept pod-to-pod within the same subnet, such as 10.128.0.1 to 10.128.0.2
so perhaps it's not the best example. fwiw, 10.128.0.0/15 would be 10.128.0.0 to 10.129.255.255, so that would work.

Author: in OCP multi nodes setup we can have also 10.130 and 10.131 I will use \14 then to widen the range in the example

        - action: Reject
          cidr: 0.0.0.0/0
``` 

<img src="{page.image('ebpf-flows-filtering/ebpf-svc-and-pods-flows.png')}" alt="eBPF Flows Filtering Kubernetes NorthSouth and EastWest Flows">

> 

Reviewer 4: You could also mention the other use-case where users would on the contrary want to have 1:1 sampling on North-South (which was what we've been asked to allow, so we know it's something users might be interested in). Something like:
If on the contrary, you're more interested in North-South traffic but still without loosing visibility on East-West, you can simply edit the last rule as:

```
        - action: Accept
          cidr: 0.0.0.0/0
          sampling: 1
```

Author: won't that be a special case of the general case I used in my example ?

Reviewer 4: If I read your example correctly, they're 2 specific and different use cases:
- your example shows keeping East-West and rejecting all North-South
- my suggestion is to show keeping both North-South and East-West, but with higher sampling defined on East-West

... unless if you consider traffic to 172.30.0.0/16 being the North-South traffic? But I don't think it's right, since Service IPs are also used in E-W (services configured as ClusterIP)

I'd rather define as "North-South" anything that has any IP neither in 10.128.0.0/14 nor in 172.30.0.0/16

For instance, a pod 10.128.0.1 that establishes a connection to an outside service like 8.8.8.8 should be categorized in N-S, right?

Author: I added additional case for 8.8.8.8 with different config options lmk if that is what u were looking for

Reviewer 4: Actually I was thinking of it as something part of the same E-W / N-S use case (it's just that I had a broader definition of N-S than what you're showing)

If I had to distinguish E-W / N-S traffic with different sampling, I'd do it that way:

```
      rules:
        - action: Accept
          cidr: 0.0.0.0/0
          sampling: 1
        - action: Accept
          cidr: 10.128.0.0/14
          peerCIDR: 10.128.0.0/14
          sampling: 50
        - action: Accept
          cidr: 172.30.0.0/16
          peerCIDR: 10.128.0.0/14
          sampling: 50
```

So that would use sampling 50 for pod-to-pod and pod-to-service, and 1 for everything else, including external-to-service or pod-to-external.

Author: this rule open things up more than 8.8.8.8 I would like to open things up only for 8.8.8.8 ping to explore the APIs options I will merge my two usecases

Reviewer 4: sure, thanks

### Filter flows with packet drops

Let's filter Kubernetes service flows that include a packet drop and discard all others.
For this use case, the `PacketDrop` feature must be enabled with eBPF in `privileged` mode,
as demonstrated in the configuration below.

```yaml
agent:
  type: eBPF
  ebpf:
    privileged: true
    features:
      - PacketDrop
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 172.30.0.0/16
          pktDrops: true

>         - action: Reject

Reviewer 4: nitpicking but I think in that case the Reject rule isn't necessary since unmatched flows are rejected by default (it might still be added on purpose, just to remove any ambiguity)

          cidr: 0.0.0.0/0
```

> <img src="{page.image('ebpf-flows-filtering/ebpf-filter-svc-pkt-drops.png')}" alt="eBPF Flows Filtering Kubernetes Services with Packet Drop">

Reviewer 2: Global comment for the use cases: For all the images of the Traffic flows table that come after the YAML, it would be good to have a brief little blurb after just drawing attention to what about the image pertains to the example use case.

```yaml
agent:
  type: eBPF
  ebpf:
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 0.0.0.0/0
          protocol: TCP
          tcpFlags: SYN
          sampling: 1
```

<img src="{page.image('ebpf-flows-filtering/ebpf-filter-with-tcpflags.png')}" alt="eBPF Flows Filtering TCP flows using TCP flags">

### Filter DNS query over ports 53 and 5353 for both TCP and UDP

This Use case involves capturing DNS flows over both `TCP` and `UDP`,
with the option to enable the `DNSTracking` feature for enhanced DNS latency insights.

```yaml
agent:
  type: eBPF
  ebpf:
    features:
      - DNSTracking
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 0.0.0.0/0
          sourcePorts: '53,5353'
          sampling: 1
```

<img src="{page.image('ebpf-flows-filtering/ebpf-filter-DNS-flows.png')}" alt="eBPF Flows Filtering DNS flows">

## Conclusion

By filtering flows at the kernel level with eBPF, we maximize efficiency,
ensuring only the most relevant data is processed and stored.
This approach is critical for scalability, cost reduction, and real-time network insights.

## Feedback

We hope you liked this article!

> Netobserv is an OpenSource project [available on github](https://github.com/netobserv).

Reviewer 4: nit: open source is 2 words

```suggestion
Netobserv is an open source project [available on github](https://github.com/netobserv).
```

## Key Use Cases

🚀 Reducing Observability Overhead – Avoid logging irrelevant flows in high-traffic Kubernetes clusters.

🔐 Security Filtering – Focus on anomalous or suspicious traffic while ignoring normal flows.

🌐 Network Performance Monitoring – Capture only high-latency or dropped-packet flows for troubleshooting.

### Filter EastWest and NorthSouth flows

Let's allow serviceIP 172.30.100.64:80 with sampling 10
and allow pods traffic between subnet 10.128.0.0/16 and 10.129.0.0/16 with sampling 20 and reject everything else

```yaml
agent:
    type: eBPF
    ebpf:
      flowFilter:
        enable: true
        rules:
        - action: Accept
          cidr: 172.30.0.0/16
          sampling: 10
        - action: Accept
          cidr: 10.128.0.0/16
          peerCIDR: 10.129.0.0/16
          sampling: 20
        - action: Reject
          cidr: 0.0.0.0/0
``` 

<img src="{page.image('ebpf-flows-filtering/ebpf-svc-and-pods-flows.png')}" alt="eBPF Flows Filtering Kubernetes NorthSouth and EastWest Flows">

### Filter flows with packet drops

let's filter any kubernetes service flows with a packet drop and reject everything else,

> Please not for this use case, we must enable `PacketDrop` feature with eBPF in `privileged` mode as shown in the below configuration

Reviewer: e missing in note

A SYN flood is a Denial-of-Service (DoS) attack where an attacker overwhelms a target system by sending a large
number of SYN packets without completing the three-way handshake,
depleting system resources and disrupting legitimate connections.

```yaml
agent:
  type: eBPF
  ebpf:
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 0.0.0.0/0
          protocol: TCP
          tcpFlags: SYN
          sampling: 1
```

<img src="{page.image('ebpf-flows-filtering/ebpf-filter-with-tcpflags.png')}" alt="eBPF Flows Filtering TCP flows using TCP flags">
<p style="text-align: center">Figure 4: eBPF Flows Filtering TCP flows using TCP flags</p>

### Filter DNS query over ports 53 and 5353 for both TCP and UDP

This Use case involves capturing DNS flows over both `TCP` and `UDP`,
with the option to enable the `DNSTracking` feature for enhanced DNS latency insights.

```yaml
agent:
  type: eBPF
  ebpf:
    features:
      - DNSTracking
    flowFilter:
      enable: true
      rules:
        - action: Accept
          cidr: 0.0.0.0/0
          sourcePorts: '53,5353'

>           sampling: 1

Reviewer 5: For this to work properly, don't you need to set DNS_TRACKING_PORT?
```
    advanced:
      env:
        DNS_TRACKING_PORT: "5353"```

Author: I didn't want to export this back door config in the blog but ofc for that u need DNS to listen to 5353

Reviewer 5: ok /lgtm
